android-client-sdk/src/main/java/com/devcycle/sdk/android/model/ReadOnlyVariable.kt

```diff
@@ -1,14 +1,94 @@
 package com.devcycle.sdk.android.model
 
-import com.fasterxml.jackson.annotation.JsonIgnoreProperties
 import com.fasterxml.jackson.annotation.JsonProperty
+import com.fasterxml.jackson.core.JsonParseException
+import com.fasterxml.jackson.core.JsonParser
+import com.fasterxml.jackson.core.TreeNode
+import com.fasterxml.jackson.databind.*
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize
+import com.fasterxml.jackson.databind.node.JsonNodeType
+import com.fasterxml.jackson.databind.node.ObjectNode
+import org.json.JSONArray
+import org.json.JSONObject
 
-@JsonIgnoreProperties(ignoreUnknown = true)
-data class ReadOnlyVariable<T>(
+class VariableDeserializer : JsonDeserializer<BaseConfigVariable>() {
+    override fun deserialize(p: JsonParser, ctxt: DeserializationContext): BaseConfigVariable {
+        val mapper = p.codec
+        val root: ObjectNode = mapper.readTree<TreeNode>(p) as ObjectNode
+        return if (root.get("value").nodeType == JsonNodeType.OBJECT) {
+            mapper.treeToValue(root, JSONObjectConfigVariable::class.java)
+        } else if (root.get("value").nodeType == JsonNodeType.ARRAY) {
+            mapper.treeToValue(root, JSONArrayConfigVariable::class.java)
+        } else if (root.get("value").nodeType == JsonNodeType.NUMBER) {
+            mapper.treeToValue(root, NumberConfigVariable::class.java);
+        } else if (root.get("value").nodeType == JsonNodeType.STRING) {
+            mapper.treeToValue(root, StringConfigVariable::class.java);
+        } else if (root.get("value").nodeType == JsonNodeType.BOOLEAN) {
+            mapper.treeToValue(root, BooleanConfigVariable::class.java)
+        } else {
+            throw JsonParseException(
+                p,
+                "Unable to parse variable value for key ${root.get("key").asText()}"
+            )
+        }
+    }
+}
+
+@JsonDeserialize(using = VariableDeserializer::class)
+abstract class BaseConfigVariable {
```

Comment:
kind of nitpicky, but I think it would make more sense to only have the `value` field be abstract since its the only one that changes with the different extension

Reply:
they all need to be declared abstract when they are declared as properties like this, otherwise the compiler complains that they need to be initialized. I'd love to have a better way to write this with less duplication etc. but I couldn't find one at the time.

Reply:
should this file be renamed to `BaseConfigVariable` though?

Reply:
should this file be renamed to `BaseConfigVariable` though?